frontend/src/routes/decrypt/+page.svelte

```diff
@@ -11,92 +12,173 @@
   import { onMount } from "svelte";
   import Auth from "./auth.svelte";
   import Password from "./password.svelte";
+  import { getError, hasError } from "$lib/form";
+  import { AlertCircle } from "lucide-svelte";
 
   onMount(() => {
     renderLoginButton(document.getElementById("login-button"));
   });
 
-  let encodedPayload = "";
-  let plainText: Promise<string> | null = null;
-  let error: any = null;
-  let decrypt: () => Promise<string>;
+  const DecryptFormSchema = z.object({
+    encodedPayload: z.string().min(1),
+    payload: z
+      .custom<Cipherly.Payload | null>()
+      .refine((payload) => {
+        // version validation
+        return true; // valid
+      }, "wrong version")
+      .refine((payload) => {
+        // validate another thing
+        return true; // valid
+      }, "something else"),
+  });
+  type DecryptFormData = z.infer<typeof DecryptFormSchema>;
+
+  let validationError: z.ZodError | undefined;
+  let formData: DecryptFormData = {
+    encodedPayload: "",
+    payload: null,
+  };
+
+  let loading = false;
+  let plainText: string | null = null;
+  let decryptionError: any = null;
+  let password = "";
 
   if (location.hash) {
-    encodedPayload = location.href;
+    formData.encodedPayload = location.href;
   }
 
-  let payload: Cipherly.Payload | null;
   $: {
-    error = null;
-    payload = null;
-    plainText = null;
     try {
-      payload =
-        encodedPayload === "" ? null : Cipherly.decodePayload(encodedPayload);
+      formData.payload = formData.encodedPayload
+        ? Cipherly.decodePayload(formData.encodedPayload)
+        : null;
+      decryptionError = null;
     } catch (err) {
-      console.log(err);
-      error = err;
+      decryptionError = err;
```

Comment:
I'd like to get rid of this in favor of the validation logic, but I'm not sure if we can get validation to run, say, when the encodedPayload field is modified?

Reply:
That should be possible I think, I'll see how we can achieve that.